### ISSUE TITLE
Support "evaluate" keyword

### ISSUE THREAD

**Issue by weiguoz** [sql-machine-learning/sqlflow#660]
Sometimes users may 
- want to evaluate a previously trained model on new data. [ref](https://github.com/sql-machine-learning/sqlflow/issues/656)
- specify evaluate dataset while executing `TRAIN`

We propose to use `EVALUATE` to fulfill these.

### Proposals

1. Evaluation-only

   ```SQL
   SELECT * FROM t
   EVALUATE model_file
   ```

2. Evaluation with training

   ```SQL
   SELECT * FROM t1 WHERE sqlflow_random < 0.8
   TRAIN DNNClassifier
   EVALUATE ON (SELECT * FROM t1 WHERE sqlflow_random >= 0.8)
   WITH
     ...
   INTO model_file
   ```

   The two above SQL statements are for advanced users.  SQLFlow also provides a simplified implementation to `TRAIN` and `EVALUATE`(a.k.a validation)

3. Implicit  evaluation with training

   ```SQL
   SELECT * FROM t
   TRAIN DNNClassifier
   WITH
   ...
   INTO model_file
   ```
   described by [training_and_validation](https://github.com/sql-machine-learning/sqlflow/blob/develop/doc/training_and_validation.md)

**Comment by terrytangyuan** [comment]
How do users specify the label columns for two tables? The label columns might be different.

**Comment by tonyyang-svail** [comment]
> How do users specify the label columns for two tables? The label columns might be different.

@terrytangyuan One may use `AS` keyword to map different column names. For example, we have two tables `table_train: c1, c2, c3, class` and `table_eval: c1, c2, c3, target`. One can write

```SQL
SELECT * FROM table_train
TRAIN DNNClassifier
EVALUATE ON (SELECT c1, c2, c3, target as label FROM table_train)
WITH
  ...
INTO model_file
```

**Comment by tonyyang-svail** [comment]
@weiguoz In solution 2, how to support arbitrary SQL statement in both train select and eval select?

**Comment by tonyyang-svail** [comment]
@typhoonzero and I discussed another possible solution several months ago. Mark this as solution 4.

Say we have our training and validation data in two tables `train` and `val` respectively.

```
mysql> select * from train;
+--------------+-------------+--------------+-------------+-------+
| sepal_length | sepal_width | petal_length | petal_width | class |
+--------------+-------------+--------------+-------------+-------+
|          4.6 |         3.1 |          1.5 |         0.2 |     0 |
...

mysql> select * from val;
+--------------+-------------+--------------+-------------+-------+
| sepal_length | sepal_width | petal_length | petal_width | class |
+--------------+-------------+--------------+-------------+-------+
|          6.3 |         2.7 |          4.9 |         1.8 |     2 |
...
```

One can combine them and add another column called `sqlflow_train_val_split`. The `train`'s `sqlflow_train_val_split` is set to `0.0` and the `val`'s `sqlflow_train_val_split` is set to `1.0`. In MySQL this can be done as

```
mysql> create table train_val select *, 1.0 as sqlflow_train_val_split from val
    -> union all
    -> select *, 0.0 as sqlflow_train_val_split from train;
Query OK, 120 rows affected (0.01 sec)
Records: 120  Duplicates: 0  Warnings: 0

mysql> select * from train_val;
+--------------+-------------+--------------+-------------+-------+-------------------------+
| sepal_length | sepal_width | petal_length | petal_width | class | sqlflow_train_val_split |
+--------------+-------------+--------------+-------------+-------+-------------------------+
|          6.3 |         2.7 |          4.9 |         1.8 |     2 |                     1.0 |
...
|          4.8 |         3.1 |          1.6 |         0.2 |     0 |                     0.0 |
...
```

We still write

```SQL
SELECT * FROM train_val
TRAIN DNNClassifier
WITH
...
INTO model
```

However, our train validation splitter detects there is a column named `sqlflow_train_val_split`, so it skips the splitting and uses `0.8` as the default threshold to construct train/val dataset.

**Comment by typhoonzero** [comment]
@tonyyang-svail I vote for the original solution of No. 2:

1. Creating a new temporary merged table takes time and storage.
2. User may need to use there own way to split train/validation dataset, it does not make sense to let them then merge them together again.

**Comment by wangkuiyi** [comment]
This discussion confused the concepts of "evaluate" and "validate".

According to https://towardsdatascience.com/train-validation-and-test-sets-72cb40cba9e7, 

- the training dataset is for fitting the model
- the test dataset is for an unbiased evaluation
- the validatation dataset is used in the training process, for hyperparameters tuning and etc, but not for fitting the model.

I am closing this issue and create two separate issues for discussions on evaluation and validation:

1. https://github.com/sql-machine-learning/sqlflow/issues/674
1. https://github.com/sql-machine-learning/sqlflow/issues/675
